### PR TITLE
Shopping feature specs

### PR DESCRIPTION
**Pull request by v-tsvid**

### CODE REVIEW

File: app/controllers/concerns/checkout_redirecting.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+module CheckoutRedirecting
+  extend ActiveSupport::Concern
+  
+  private
+    def redirect_if_wrong_step(checkout_form, step)
+      order_next_step = checkout_form.model.next_step
+      notice = wrong_step_notice(order_next_step, step)
+      
+      if notice
```
Comment (yamnich) [inline app/controllers/concerns/checkout_redirecting.rb]: guard clause

File: app/controllers/concerns/cookies_handling.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    def push_to_cookies(order_items)
+      order_items << OrderItem.new(book_id: params[:book_id], 
+                                   quantity: params[:quantity])
+    end
+
+    def pop_from_cookies(order_items)
+      order_items.each do |item|
+        order_items.delete(item) if item.book_id.to_s == params[:book_id]
```
Comment (yamnich) [inline app/controllers/concerns/cookies_handling.rb]: order_items.where(book_id: params[:book_id].to_i).delete_all    ?

File: app/controllers/concerns/cookies_handling.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    def write_to_cookies(items)
+      cookies[:order_items] = ''
+      items.each do |item|
+        cookies[:order_items] = { 
+          value: [
+            cookies[:order_items],
+            item[:book_id],
+            item[:quantity]].join(' '),
+          expires: 30.days.from_now }
```
Comment (yamnich) [inline app/controllers/concerns/cookies_handling.rb]: simplify

File: app/controllers/concerns/cookies_handling.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      end
+    end
+
+    def read_from_cookies
+      order_items = Array.new
+      if cookies[:order_items]
```
Comment (yamnich) [inline app/controllers/concerns/cookies_handling.rb]: return order_items unless cookies[:order_items]

File: app/controllers/concerns/cookies_handling.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        cookies[:order_items].split(' ').
+          partition.with_index{ |v, index| index.even? }.transpose.each do |item|
```
Comment (yamnich) [inline app/controllers/concerns/cookies_handling.rb]: move to method

File: app/controllers/concerns/order_helpers.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+module OrderHelpers
+  extend ActiveSupport::Concern
+  
+  def current_order
+    current_customer ? current_customer.current_order_of_customer : nil
+  end
+
+  def last_order
+    current_order || (current_customer ? Order.where(
```
Comment (yamnich) [inline app/controllers/concerns/order_helpers.rb]: scope

File: app/controllers/customers/omniauth_callbacks_controller.rb [hunk 7]
```diff
@@ -1,55 +1,17 @@
 class Customers::OmniauthCallbacksController < Devise::OmniauthCallbacksController
   def facebook
-    @customer = Customer.from_omniauth(request.env["omniauth.auth"])
-
-    @customer.email = generated_email(@customer) if !@customer.email
-     
-
-    if @customer.save
+    # @customer = Customer.from_omniauth(request.env["omniauth.auth"])
```
Comment (yamnich) [inline app/controllers/customers/omniauth_callbacks_controller.rb]: aaa

File: app/controllers/customers/omniauth_callbacks_controller.rb [hunk 8]
```diff
@@ -1,55 +1,17 @@
...
+    if @customer && @customer.save
       sign_in_and_redirect @customer, event: :authentication 
       set_flash_message(:notice, :success, 
                         kind: "Facebook") if is_navigational_format?
     else
       session["devise.facebook_data"] = request.env["omniauth.auth"] 
-      redirect_to root_path
+      redirect_to new_customer_registration_path
       set_flash_message(:notice, :failure, kind: "Facebook", 
         reason: "#{@customer.inspect}") if is_navigational_format?
```
Comment (yamnich) [inline app/controllers/customers/omniauth_callbacks_controller.rb]: do not use inspect when message is displayed

File: app/controllers/order_items_controller.rb [hunk 9]
```diff
@@ -3,57 +3,53 @@ class OrderItemsController < ApplicationController
   authorize_resource
 
   def index
-    @order = order_with_items
+    @order = OrderFiller.new(get_order).add_items_to_order(read_from_cookies)
+    cookies.delete('order_items') if @order.persisted?
+
     @order_items = @order.order_items
 
     if @order_items.empty?
```
Comment (yamnich) [inline app/controllers/order_items_controller.rb]: guard clause

File: app/controllers/orders_controller.rb [hunk 10]
```diff
@@ -11,18 +11,19 @@ def show
 
   def update
```
Comment (yamnich) [inline app/controllers/orders_controller.rb]: should be one action update

File: app/helpers/orders_helper.rb [hunk 11]
```diff
@@ -1,7 +1,6 @@
 module OrdersHelper
 
-  def order_completed_date(order)
-    DateDecorator.decorate(order.completed_date).date || 
-    t("orders_page.not_completed")
+  def actual_orders(orders, state)
+    orders.where(state: state).reverse_order
```
Comment (yamnich) [inline app/helpers/orders_helper.rb]: scope

File: app/controllers/concerns/cookies_handling.rb (hunk 2 above)
Comment (v-tsvid) [inline app/controllers/concerns/cookies_handling.rb]: nope. cause order_items is an array

File: app/controllers/concerns/cookies_handling.rb (hunk 4 above)
Comment (v-tsvid) [inline app/controllers/concerns/cookies_handling.rb]: why? i need to return order_items despite of cookies[:order_items] is nil or not.

File: app/controllers/orders_controller.rb (hunk 10 above)
Comment (v-tsvid) [inline app/controllers/orders_controller.rb]: discussed and decided to leave it alone

**Pull request merged**